File: watch/wearos/src/main/java/app/vescape/wear/StatusLayouts.kt
package app.vescape.wear

import androidx.compose.foundation.layout.Box
import androidx.compose.foundation.layout.Column
import androidx.compose.foundation.layout.Spacer
import androidx.compose.foundation.layout.fillMaxSize
import androidx.compose.foundation.layout.height
import androidx.compose.foundation.layout.padding
import androidx.compose.runtime.Composable
import androidx.compose.runtime.getValue
import androidx.compose.ui.Alignment
import androidx.compose.ui.Modifier
import androidx.compose.ui.text.style.TextAlign
import androidx.compose.ui.unit.dp
import androidx.wear.compose.material.CircularProgressIndicator
import androidx.wear.compose.material.MaterialTheme
import androidx.wear.compose.material.Text

/**
 * No fresh frames: name the reason from the watch-local [PhoneLink] instead of an anonymous
 * spinner, so "Bluetooth to phone is down" and "phone app missing" and "app just isn't riding"
 * are distinguishable at a glance.
 */
@Composable
internal fun DisconnectedLayout(ambient: AmbientMode) {
    val link by TelemetryState.phoneLink
    val (title, caption) = when (link) {
        PhoneLink.UNKNOWN -> "Connecting…" to ""
        PhoneLink.NO_PHONE -> "No phone link" to "Check Bluetooth"
        PhoneLink.PHONE_ONLY -> "Phone linked" to "Vescape app not found"
        PhoneLink.APP_REACHABLE -> "Phone connected" to "No board session"
    }
    // Ambient keeps the reason on screen — it is exactly the state a rider needs while not looking
    // — but never the spinner: a permanent animation is the one thing an always-on panel cannot pay
    // for.
    StatusLayout(
        title = title,
        caption = caption,
        spin = !ambient.active && link != PhoneLink.NO_PHONE,
        ambient = ambient,
    )
}

@Composable
private fun StatusLayout(title: String, caption: String, spin: Boolean, ambient: AmbientMode) {
    Box(modifier = Modifier.fillMaxSize(), contentAlignment = Alignment.Center) {
        if (spin) {
            CircularProgressIndicator(
                modifier = Modifier.fillMaxSize().padding(8.dp),
                indicatorColor = SpeedColor,
                trackColor = GuideColor,
                strokeWidth = 4.dp,
            )
        }
        Column(
            modifier = Modifier.padding(horizontal = 24.dp),
            horizontalAlignment = Alignment.CenterHorizontally,
        ) {
            Text(
                text = title,
                style = MaterialTheme.typography.title3,
                color = ambient.readout(PrimaryText),
                textAlign = TextAlign.Center,
            )
            if (caption.isNotEmpty()) {
                Spacer(modifier = Modifier.height(4.dp))
                Text(
                    text = caption,
                    style = MaterialTheme.typography.caption2,
                    color = ambient.skeleton(SecondaryText),
                    textAlign = TextAlign.Center,
                )
            }
        }
    }
}

File: watch/watchos/StatusLayouts.swift
import SwiftUI

/// No fresh frames: name the reason from the watch-local link view, so a dead Bluetooth link, a
/// missing phone app and an idle one are distinguishable at a glance.
///
/// Each title names the thing that is missing rather than the state of the chain, in the rider's
/// words and not the enum's: the failure is always one broken step in watch → phone → board, and
/// the last of those is not about the phone at all.
///
/// This draws inside the rim gauges, not instead of them. The shell drops its speed and duty heroes
/// while disconnected, so the reason owns the centre of the screen rather than dodging two dashes.
///
/// @parity /watch/wearos/src/main/java/app/vescape/wear/StatusLayouts.kt `DisconnectedLayout`
struct DisconnectedLayout: View {
  let link: MirrorPhoneLink
  let ambient: AmbientMode

  var body: some View {
    let (title, caption) = message

    VStack(spacing: 4) {
      Text(title)
        .font(.system(size: 15, weight: .semibold))
        .foregroundStyle(ambient.readout(Palette.primaryText))
        .multilineTextAlignment(.center)
      if !caption.isEmpty {
        Text(caption)
          .font(.system(size: 12))
          .foregroundStyle(ambient.skeleton(Palette.secondaryText))
          .multilineTextAlignment(.center)
      }
    }
    .padding(.horizontal, Rim.innerInset + 8)
  }

  private var message: (String, String) {
    switch link {
    case .unknown:
      return ("Connecting…", "")
    // The title already says it. "Check Bluetooth" was the only fix a rider could try, and they try
    // it without being told.
    case .noPhone:
      return ("Phone not connected", "")
    // The companion reads as absent when the app is missing *or* too old, so the caption has to
    // cover both; naming only one of them sends half the riders down the wrong path.
    case .phoneOnly:
      return ("Phone app missing", "Install or update Vescape")
    case .appReachable:
      return ("Board not connected", "Connect it on your phone")
    }
  }
}

/// A page whose content belongs to a later slice of the watchOS port. It exists so the page
/// structure is Android's from the start — the crown and the swipes travel the same axes, in the
/// same order — rather than being re-shuffled under the rider as each channel lands.
struct PendingPage: View {
  let title: String

  var body: some View {
    VStack(spacing: 4) {
      Text(title)
        .font(.system(size: 15, weight: .semibold))
        .foregroundStyle(Palette.secondaryText)
      Text("Not on the wrist yet")
        .font(.system(size: 11))
        .foregroundStyle(Palette.dimText)
    }
    .multilineTextAlignment(.center)
    .padding(.horizontal, Rim.innerInset)
  }
}
